### ISSUE TITLE
404 errors on 'missing' tiles

### ISSUE THREAD

**Issue by jnicho02** [consbio/mbgl-renderer#37]
This may well be nothing to do with mbgl-renderer..but maybe you've seen similar (and dealt with it)?

I am hosting some custom .pbf vector tiles on Amazon S3. Data is sparse, so there is no .pbf for many gridrefs. When I run MapboxGL to make an online map it is quite happy with this. When I run mbgl-renderer I get a 404 error for the non-existent .pbf

```
Error with request for: http://nnnnnn.s3-website-eu-west-1.amazonaws.com/wind/20191025/13/3978/2766.pbf
status: 404
Error rendering map
```

**Comment by jnicho02** [comment]
Ok, so not mbgl-renderer's problem/fault. The vector tileserver should be returning an empty .pbf. I have achieved this by using CloudFront and returning 0/0/0.pbf and return code 200, although I may want to construct an empty pbf instead.

MapboxGL in a web page does not stop working, but it does still register an error behind the scenes.

**Comment by brendan-ward** [comment]
@jnicho02 we didn't specifically handle 404 errors here:
https://github.com/consbio/mbgl-renderer/blob/master/src/render.js#L246-L250

However, it shouldn't be hard to catch that case specifically, and log a warning instead of failing outright.  The outcome would be the same (tiles would not be rendered in output), without requiring changes to the origin server.

Do you have a style JSON file you'd be willing to share, so that I can think about how we would approach this from a testing perspective?

**Comment by jnicho02** [comment]
Basic workflow is:
1. get a geojson object
2. use tippecanoe to create a .pbf
3. use github.com/mapbox/mbutil to unpack it to individual .pbfs
4. repoint the metadata json at the new tile url root
5. copy to S3 giving .pbf a content-type application/x-protobuf and .json application/json

For testing purposes i'm just wondering whether you could run a web server locally (like node's http-server).

There is a chance that AWS S3 throws a 403 and/or a 404. I ended up repointing both in CloudFront to 200

Give me a few days and i'll see if I can run locally....or reduce the S3 requirements to a minimum (e.g. does it need a sprite sheet or can we just ignore it?)

**Comment by brendan-ward** [comment]
@jnicho02 thanks for sharing more information!

I think it is entirely possible to host locally using whatever static server you prefer; you just need to serve a directly with the appropriate MIME types and CORS settings.  

I created a tiny one in Go [long ago](https://github.com/consbio/go-http-fileserver) if you are looking for something extremely minimal in that space (though I've not tested with .pbfs, so no assurances about correct MIME type there).  Otherwise http-server, Express, Caddy, NGINX, etc could all do this.  The only catch is that your style files need to point to local URLs rather than remote.

Per what we want to solve in `mbgl-renderer` 404s seem reasonable for us to skip with warning to user; not sure about 401s / 403s - we may or may not want to fail harder in those cases.

Presumably you've also seen some of the other Mapbox utils for shipping from local tilesets to S3?  If not:
* [https://github.com/mapbox/node-mbtiles](https://github.com/mapbox/node-mbtiles)
* [https://github.com/mapbox/mapbox-tile-copy](https://github.com/mapbox/mapbox-tile-copy)

**Comment by brendan-ward** [comment]
Resolved by 91f3ec2defb4596574e353a0bbeae9101696b982
